File: data_source_service/app/controllers/user_controller.rb
class UserController < ApplicationController
  def get_all_users
    all_users = UserUseCases.get_all_users

    render json: all_users, status: 200
  end

  def get_user
    begin
      id = params[:id]

      user = UserUseCases.get_user id

      render json: user, status: 200
    rescue StandardError
      render status: 404
    end
  end

  def create_user
    begin
      serialize_request_body = JSON.parse request.raw_post

      id = serialize_request_body[UserConstants::USER_MAP_ID_KEY]
      username = serialize_request_body[UserConstants::USER_MAP_USERNAME_KEY]

      UserUseCases.create_user username, id

      render status: 201
    rescue StandardError
      render status: 409
    end
  end

  def update_user
    begin
      serialize_request_body = JSON.parse request.raw_post

      id = params[:id]
      username = serialize_request_body[UserConstants::USER_MAP_USERNAME_KEY]

      UserUseCases.update_user username, id

      render status: 200
    rescue StandardError
      render status: 404
    end
  end

  def delete_user
    begin
      id = params[:id]

      UserUseCases.delete_user id

      render status: 204
    rescue StandardError
      render status: 404
    end
  end
end

File: in_memory_cache_service/src/controllers/user_controller.ts
import { Request, Response } from "express";
import UserRepositoryFactory from "../domains/user/user_repository_factory";
import { StatusCodes } from "http-status-codes";
import UserEntityFactory from "../domains/user/user_entity_factory";

class UserController {
  public static async getAllUsers(
    request: Request,
    response: Response,
  ): Promise<void> {
    const userRepository = await UserRepositoryFactory.getInstance();

    const users = await userRepository.getUsers();

    response.json(users);
  }

  public static async getUser(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const userId = Number(request.params.id);

      const userRepository = await UserRepositoryFactory.getInstance();

      const users = await userRepository.getUser(userId);

      response.json(users);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }

  public static async createUser(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const userId = Number(request.body.id);
      const userUsername = request.body.username;

      const userRepository = await UserRepositoryFactory.getInstance();

      const user = UserEntityFactory.getInstance(userId, userUsername);

      await userRepository.createUser(user);

      response.status(StatusCodes.CREATED).send("");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.CONFLICT).send("");
    }
  }

  public static async updateUser(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const userId = Number(request.body.id);
      const userUsername = request.body.username;

      const userRepository = await UserRepositoryFactory.getInstance();

      const user = UserEntityFactory.getInstance(userId, userUsername);

      await userRepository.updateUser(user);

      response.status(StatusCodes.OK).send("");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }

  public static async deleteUser(
    request: Request,
    response: Response,
  ): Promise<void> {
    try {
      const userId = Number(request.params.id);

      const userRepository = await UserRepositoryFactory.getInstance();

      await userRepository.deleteUser(userId);

      response.status(StatusCodes.NO_CONTENT).send("");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      response.status(StatusCodes.NOT_FOUND).send("");
    }
  }
}

export default UserController;
